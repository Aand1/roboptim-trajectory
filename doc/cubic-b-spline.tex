\documentclass {article}

\usepackage {amsmath}
\usepackage {amsfonts} % pour les lettres maths creuses \mathbb
\newcommand\real{\mathbb{R}}

\title {Computation of non uniform cubic B splines}
\author {Florent Lamiraux, LAAS-CNRS}

\begin{document}
\maketitle
\section {Definition}

\paragraph {Knots} $m \leq 4$ increasing real values:
$$
t_0 \leq t_1 \leq \cdots \leq t_{m-1} 
$$
\paragraph {Control points} $m-4$ points:
$$
\mathbf{P}_{i} \in \mathbb{R}^d, 0\leq i \leq m-5.
$$
\paragraph {Cubic B spline} parametric curve defined by
\begin{equation}\label{eq:spline-definition}
B(t) = \sum_{i=0}^{m-5} b_{i,3} (t) P_i,\ \ t\in [t_3,t_{m-4}]
\end{equation}

\section {Computation of basis functions}

\subsubsection* {Basis functions of order 0}

\begin{equation}\label {eq:basis-0}
b_{j,0}(t) = \mathbb{I}_{[t_j, t_{j+1}]}(t) = \left\{ \begin{matrix} 1 & \mathrm{if} \quad t_j \leq t < t_{j+1} \\ 0 & \mathrm{otherwise} \end{matrix} \right.,\qquad j=0,\ldots, m-2
\end{equation}

\subsubsection* {Recursion formula}

\begin{equation}\label {eq:recursion}
b_{j,n}(t) := \frac{t - t_j}{t_{j+n} - t_j} b_{j,n-1}(t) + \frac{t_{j+n+1} - t}{t_{j+n+1} - t_{j+1}} b_{j+1,n-1}(t) ,\qquad j=0,\ldots, m-n-2. 
\end{equation}

\subsubsection* {Basis functions of order 1}

\begin{equation}\label {eq:basis-1}
b_{j,1}(t) = \frac{t - t_j}{t_{j+1} - t_j} \mathbb{I}_{[t_j, t_{j+1}]}(t) + \frac{t_{j+2} - t}{t_{j+2} - t_{j+1}} \mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) ,\qquad j=0,\ldots, m-3
\end{equation}

\subsubsection* {Basis functions of order 2}

\begin{eqnarray*}
b_{j,2}(t) &=& \frac{t - t_j}{t_{j+2} - t_j} \left(\frac{t - t_j}{t_{j+1} - t_j} \mathbb{I}_{[t_j, t_{j+1}]}(t) + \frac{t_{j+2} - t}{t_{j+2} - t_{j+1}} \mathbb{I}_{[t_{j+1}, t_{j+2}]}(t)\right)\\
&&+ \frac{t_{j+3} - t}{t_{j+3} - t_{j+1}}\left(\frac{t - t_{j+1}}{t_{j+2} - t_{j+1}} \mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) + \frac{t_{j+3} - t}{t_{j+3} - t_{j+2}} \mathbb{I}_{[t_{j+2}, t_{j+3}]}(t)\right)\\
&=&\frac{(t - t_j)^2}{(t_{j+2} - t_j)(t_{j+1} - t_j)}\mathbb{I}_{[t_j, t_{j+1}]}(t) \\
&& + \left(\frac{t - t_j}{t_{j+2} - t_j}\frac{t_{j+2} - t}{t_{j+2} - t_{j+1}}+
\frac{t_{j+3} - t}{t_{j+3} - t_{j+1}}\frac{t - t_{j+1}}{t_{j+2} - t_{j+1}}\right)
\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
&&+\frac{(t_{j+3} - t)^2}{(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})} \mathbb{I}_{[t_{j+2}, t_{j+3}]}(t)\\
&& j=0,\ldots, m-4. 
\end{eqnarray*}

\subsubsection* {Basis functions of order 3}

\begin{eqnarray*}
b_{j,3}(t) &=& \frac{t - t_j}{t_{j+3} - t_j} 
\left(\begin{array}{l}
\frac{(t - t_j)^2}{(t_{j+2} - t_j)(t_{j+1} - t_j)}\mathbb{I}_{[t_j, t_{j+1}]}(t) \\
+ (\frac{t - t_j}{t_{j+2} - t_j}\frac{t_{j+2} - t}{t_{j+2} - t_{j+1}}+
\frac{t_{j+3} - t}{t_{j+3} - t_{j+1}}\frac{t - t_{j+1}}{t_{j+2} - t_{j+1}})
\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
+\frac{(t_{j+3} - t)^2}{(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})} \mathbb{I}_{[t_{j+2}, t_{j+3}]}(t)
\end{array}\right) \\
&&+ \frac{t_{j+4} - t}{t_{j+4} - t_{j+1}}
\left(\begin{array}{l}\frac{(t - t_{j+1})^2}{(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})}\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
+ (\frac{t - t_{j+1}}{t_{j+3} - t_{j+1}}\frac{t_{j+3} - t}{t_{j+3} - t_{j+2}}+
\frac{t_{j+4} - t}{t_{j+4} - t_{j+2}}\frac{t - t_{j+2}}{t_{j+3} - t_{j+2}})
\mathbb{I}_{[t_{j+2}, t_{j+3}]}(t) \\
+\frac{(t_{j+4} - t)^2}{(t_{j+4} - t_{j+2})(t_{j+4} - t_{j+3})} \mathbb{I}_{[t_{j+3}, t_{j+4}]}(t)
\end{array}\right)\\
&=& \frac{(t - t_j)^3}{(t_{j+3} - t_j)(t_{j+2} - t_j)(t_{j+1} - t_j)}\mathbb{I}_{[t_j, t_{j+1}]}(t) \\
&&+ (\frac{t - t_j}{t_{j+3} - t_j})(\frac{(t - t_j)(t_{j+2} - t)}{(t_{j+2} - t_{j+1})(t_{j+2} - t_j)}+
\frac{(t_{j+3} - t)(t - t_{j+1})}{(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})})
\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
&&+\frac{(t - t_j)(t_{j+3} - t)^2}{(t_{j+3} - t_j)(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})} \mathbb{I}_{[t_{j+2}, t_{j+3}]}(t)\\
&&+\frac{(t_{j+4} - t)(t - t_{j+1})^2}{(t_{j+4} - t_{j+1})(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})}\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
&&+(\frac{t_{j+4} - t}{t_{j+4} - t_{j+1}})(\frac{(t - t_{j+1})(t_{j+3} - t)}{(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})}+
\frac{(t_{j+4} - t)(t - t_{j+2})}{(t_{j+4} - t_{j+2})(t_{j+3} - t_{j+2})})
\mathbb{I}_{[t_{j+2}, t_{j+3}]}(t) \\
&&+\frac{(t_{j+4} - t)^3}{(t_{j+4} - t_{j+1})(t_{j+4} - t_{j+2})(t_{j+4} - t_{j+3})} \mathbb{I}_{[t_{j+3}, t_{j+4}]}(t)\\
&&j=0,\ldots, m-5
\end{eqnarray*}
\begin{eqnarray*}
b_{j,3}(t) &=& \frac{(t - t_j)^3}{(t_{j+3} - t_j)(t_{j+2} - t_j)(t_{j+1} - t_j)}\mathbb{I}_{[t_j, t_{j+1}]}(t) \\
&&+(\frac{(t - t_j)^2(t_{j+2} - t)}{(t_{j+3} - t_j)(t_{j+2} - t_{j+1})(t_{j+2} - t_j)}+
\frac{(t - t_j)(t_{j+3} - t)(t - t_{j+1})}{(t_{j+3} - t_j)(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})})
\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
&&+\frac{(t - t_j)(t_{j+3} - t)^2}{(t_{j+3} - t_j)(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})} \mathbb{I}_{[t_{j+2}, t_{j+3}]}(t)\\
&&+\frac{(t_{j+4} - t)(t - t_{j+1})^2}{(t_{j+4} - t_{j+1})(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})}\mathbb{I}_{[t_{j+1}, t_{j+2}]}(t) \\
&&+(\frac{(t_{j+4} - t)(t - t_{j+1})(t_{j+3} - t)}{(t_{j+4} - t_{j+1})(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})}+
\frac{(t_{j+4} - t)^2(t - t_{j+2})}{(t_{j+4} - t_{j+1})(t_{j+4} - t_{j+2})(t_{j+3} - t_{j+2})})
\mathbb{I}_{[t_{j+2}, t_{j+3}]}(t) \\
&&+\frac{(t_{j+4} - t)^3}{(t_{j+4} - t_{j+1})(t_{j+4} - t_{j+2})(t_{j+4} - t_{j+3})} \mathbb{I}_{[t_{j+3}, t_{j+4}]}(t)\\
&=& (B_{j, j} \mathbb{I}_{[t_j, t_{j+1}]} + B_{j,j+1} \mathbb{I}_{[t_{j+1}, t_{j+2}]} +B_{j,j+2} \mathbb{I}_{[t_{j+2}, t_{j+3}]}+B_{j,j+3} \mathbb{I}_{[t_{j+3}, t_{j+4}]})(t)\\
&&j=0,\ldots, m-5
\end{eqnarray*}
with
\begin{eqnarray*}
B_{j, j} (t) &=& \frac{(t - t_j)^3}{(t_{j+3} - t_j)(t_{j+2} - t_j)(t_{j+1} - t_j)} \\
B_{j,j+1} (t) &=& \frac{(t - t_j)^2(t_{j+2} - t)}{(t_{j+3} - t_j)(t_{j+2} - t_{j+1})(t_{j+2} - t_j)}+\frac{(t - t_j)(t_{j+3} - t)(t - t_{j+1})}{(t_{j+3} - t_j)(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})}\\
&&+\frac{(t_{j+4} - t)(t - t_{j+1})^2}{(t_{j+4} - t_{j+1})(t_{j+3} - t_{j+1})(t_{j+2} - t_{j+1})} \\
B_{j,j+2} (t) &=& \frac{(t - t_j)(t_{j+3} - t)^2}{(t_{j+3} - t_j)(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})}+\frac{(t_{j+4} - t)(t - t_{j+1})(t_{j+3} - t)}{(t_{j+4} - t_{j+1})(t_{j+3} - t_{j+1})(t_{j+3} - t_{j+2})}\\
&&+\frac{(t_{j+4} - t)^2(t - t_{j+2})}{(t_{j+4} - t_{j+1})(t_{j+4} - t_{j+2})(t_{j+3} - t_{j+2})}\\
B_{j,j+3} (t) &=& \frac{(t_{j+4} - t)^3}{(t_{j+4} - t_{j+1})(t_{j+4} - t_{j+2})(t_{j+4} - t_{j+3})}
\end{eqnarray*}

\section {Evaluation of cubic B-spline}

By substituting the expression of $b_{j,3}$ obtained in the previous section in Equation~(\ref{eq:spline-definition}), we get for $t\in [t_3,t_{m-4}]$
\begin{eqnarray*}
B(t) &=& \sum_{i=0}^{m-5} (B_{i, i} \mathbb{I}_{[t_i, t_{i+1}]} + B_{i,i+1} \mathbb{I}_{[t_{i+1}, t_{i+2}]} +B_{i,i+2} \mathbb{I}_{[t_{i+2}, t_{i+3}]}+B_{i,i+3} \mathbb{I}_{[t_{i+3}, t_{i+4}]})(t) P_i.\\
&=& \sum_{i=0}^{m-5} \sum_{j=i}^{i+3} B_{i, j} \mathbb{I}_{[t_{j}, t_{j+1}]} (t) P_i\\
&=& \sum_{j=0}^{m-2} \sum_{i=\max(0,j-3)}^{\min(j,m-5)} B_{i, j} \mathbb{I}_{[t_{j}, t_{j+1}]} (t) P_i
\end{eqnarray*}
Let $k$ be such that $t\in[t_{k},t_{k+1})$. For $t\in[t_3,t_{m-4})$, $3\leq k \leq m-5$ and
\begin{eqnarray}\label{eq:spline-evaluation}
B(t) &=& \sum_{i=k-3}^{k} B_{i, k}(t)  P_i
\end{eqnarray}
\end{document}
