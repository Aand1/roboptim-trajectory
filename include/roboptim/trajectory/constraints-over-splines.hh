#ifndef ROBOPTIM_TRAJECTORY_CONSTRAINTS_OVER_SPLINES_HH
# define ROBOPTIM_TRAJECTORY_CONSTRAINTS_OVER_SPLINES_HH

# include <vector>

# include <boost/shared_ptr.hpp>

# include <roboptim/core/differentiable-function.hh>

namespace roboptim
{
  namespace trajectory
  {
    /// \brief Constraint function on a spline's interval
    ///
    /// \tparam T Matrix type
    /// \tparam S Spline type
    template <typename T, typename S>
    class ConstraintsOverSplines
    : public roboptim::GenericDifferentiableFunction<T>
    {
    public:
      ROBOPTIM_DIFFERENTIABLE_FUNCTION_FWD_TYPEDEFS_
        (roboptim::GenericDifferentiableFunction<T>);

      typedef typename S::polynomial_t polynomial_t;
      typedef Function::interval_t interval_t;
      typedef std::vector<polynomial_t> polynomials_t;
      typedef S spline_t;
      typedef boost::shared_ptr<spline_t> splinePtr_t;
      typedef std::vector<boost::shared_ptr<S> > splines_t;

      /// \brief Constructor.
      ///
      /// \param splines vector of splines.
      /// \param splineIdx index of the spline to constrain.
      /// \param order derivation order of the spline to constain.
      /// \param startingPoint starting point of the constraint.
      /// \param inputSize input size of the problem.
      ConstraintsOverSplines (const splines_t& splines, size_t splineIdx,
                              unsigned int order, value_type startingPoint,
                              size_type inputSize);

    protected:

      /// \brief Generate the name of the constraint.
      /// \param splines vector of splines.
      /// \param splineIdx index of the spline to constrain.
      /// \param order derivation order of the spline to constain.
      /// \param startingPoint starting point of the constraint.
      static std::string GenerateName (const splines_t& splines,
                                       size_t splineIdx,
                                       unsigned int order,
                                       value_type startingPoint);

      /// \brief Compute the range of the time interval.
      /// \param splines vector of splines.
      /// \param splineIdx index of the spline to constrain.
      /// \param startingPoint starting point of the constraint.
      static interval_t ComputeInterval (const splines_t& splines,
                                         size_t splineIdx,
                                         value_type startingPoint);

      /// \brief Determine the interval index of the constraint.
      /// \param splines vector of splines.
      /// \param splineIdx index of the spline to constrain.
      /// \param startingPoint starting point of the constraint.
      static size_t ComputeIntervalIdx (const splines_t& splines,
                                        size_t splineIdx,
                                        value_type startingPoint);

      /// \brief Determine the first index in the argument vector, assuming all
      /// joint control points are contiguous.
      /// \param splines vector of splines.
      /// \param splineIdx index of the spline to constrain.
      static size_type ComputeStartIdx (const splines_t& splines,
                                        size_t splineIdx);

      polynomial_t toPoly (const_argument_ref x) const;

      virtual void impl_compute (result_ref result,
                                 const_argument_ref x) const;

      virtual void impl_gradient (gradient_ref grad,
                                  const_argument_ref x,
                                  size_type i) const;

      /// \brief Order of the spline (e.g. 3 for cubic).
      size_t order_;

      /// \brief Time range of the constraint.
      typename S::interval_t interval_;

      /// \brief Index of the influencing control points.
      size_type startingIndex_;

      /// \brief Basis polynomials used for gradient computation.
      polynomials_t basisPolynomials_;

      /// \brief Constrained spline.
      mutable spline_t spline_;

      /// \brief Index of the constrained time interval.
      size_t intervalIdx_;
    };
  }
}

#include <roboptim/trajectory/constraints-over-splines.hxx>

#endif //! ROBOPTIM_TRAJECTORY_CONSTRAINTS_OVER_SPLINES_HH
