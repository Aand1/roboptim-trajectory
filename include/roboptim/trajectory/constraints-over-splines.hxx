#ifndef ROBOPTIM_TRAJECTORY_CONSTRAINTS_OVER_SPLINES_HXX
# define ROBOPTIM_TRAJECTORY_CONSTRAINTS_OVER_SPLINES_HXX

# include <boost/format.hpp>

namespace roboptim
{
  namespace trajectory
  {
    template <typename T, typename S>
    ConstraintsOverSplines<T, S>::ConstraintsOverSplines
    (const splines_t& splines, size_t splineIdx, unsigned int order,
     value_type startingPoint, size_type inputSize)
    : GenericDifferentiableFunction<T>
        (inputSize, 2, GenerateName (splines, splineIdx,
                                     order, startingPoint)),
      dimension_ (static_cast<size_t> (splines[splineIdx]->dimension ())),
      interval_ (ComputeInterval (splines, splineIdx, startingPoint)),
      startingIndex_ (ComputeStartIdx (splines, splineIdx)),
      p_ (),
      coefs_ ()
    {
      size_t interval_idx = ComputeIntervalIdx (splines, splineIdx, startingPoint);

      for (size_t j = 0; j < dimension_ + 1; ++j)
      {
        polynomial_t p;
        switch (order)
        {
          case 1 :
          p = splines[splineIdx]->basisPolynomials()[interval_idx + j][dimension_ - j].template derivative<1>();
            break;
          case 2 :
          p = splines[splineIdx]->basisPolynomials()[interval_idx + j][dimension_ - j].template derivative<2>();
            break;
          default :
          p = splines[splineIdx]->basisPolynomials()[interval_idx + j][dimension_ - j];
            break;
        }
        coefs_.push_back(p);
      }
    }

    template <typename T, typename S>
    std::string ConstraintsOverSplines<T, S>::GenerateName
    (const splines_t& splines, size_t splineIdx, unsigned int order,
     value_type startingPoint)
    {
      std::string name;
      interval_t range = ComputeInterval (splines, splineIdx, startingPoint);

      if (order == 0)
        name = (boost::format ("f(t) / t ∈ [%.2f,%.2f] (%s)")
            % range.first % range.second
            % splines[splineIdx]->getName ()).str ();
      else
        name = (boost::format ("f^(%i)(t) / t ∈ [%.2f,%.2f] (%s)")
            % order % range.first % range.second
            % splines[splineIdx]->getName ()).str ();

      return name;
    }

    template <typename T, typename S>
    typename ConstraintsOverSplines<T, S>::size_type
    ConstraintsOverSplines<T, S>::ComputeStartIdx (const splines_t& splines,
                                                   size_t splineIdx)
    {
      size_type start_idx = 0;
      for (size_t i = 0; i < splineIdx; ++i)
        start_idx += static_cast<size_type>
                       (splines[i]->getNumberControlPoints ());
      return start_idx;
    }

    template <typename T, typename S>
    size_t ConstraintsOverSplines<T, S>::ComputeIntervalIdx
    (const splines_t& splines, size_t splineIdx, value_type startingPoint)
    {
      size_t dimension = static_cast<size_t> (splines[splineIdx]->dimension ());
      size_t interval = static_cast<size_t>
        (splines[splineIdx]->interval (startingPoint));

      return interval - dimension;
    }

    template <typename T, typename S>
    typename ConstraintsOverSplines<T, S>::interval_t
    ConstraintsOverSplines<T, S>::ComputeInterval
    (const splines_t& splines, size_t splineIdx, value_type startingPoint)
    {
      size_t dimension = static_cast<size_t> (splines[splineIdx]->dimension ());
      size_t interval_idx = ComputeIntervalIdx (splines, splineIdx,
                                                startingPoint);

      size_type idx = static_cast<size_type> (interval_idx + dimension + 1);

      return S::makeInterval (startingPoint,
                              splines[splineIdx]->knotVector ()[idx]);
    }

    template <typename T, typename S>
    void ConstraintsOverSplines<T, S>::update (const_argument_ref x) const
    {
      p_.coefs().setZero();
      for (size_t i = 0; i < dimension_ + 1; ++i)
        p_ += coefs_[i] * x[static_cast<size_type> (i)];
    }

    template <typename T, typename S>
    void ConstraintsOverSplines<T, S>::impl_compute (result_ref result, const_argument_ref x) const
    {
      this->update (x.middleRows (static_cast<long> (startingIndex_),
                                  static_cast<long> (dimension_ + 1)));
      result[0] = p_.min(interval_).second;
      result[1] = p_.max(interval_).second;
    }

    template <typename T, typename S>
    void ConstraintsOverSplines<T, S>::impl_gradient
    (gradient_ref grad, const_argument_ref x, size_type functionId) const
    {
      update (x.middleRows (startingIndex_,
                            static_cast<size_type> (dimension_ + 1)));

      if (functionId == 0)
      {
        value_type tmin_ = p_.min(interval_).first;
        for (size_t i = 0; i < dimension_ + 1; ++i)
          grad.coeffRef (startingIndex_ + static_cast<size_type> (i))
            = coefs_[i](tmin_);
      }
      else
      {
        value_type tmax_ = p_.max(interval_).first;
        for (size_t i = 0; i < dimension_ + 1; ++i)
          grad.coeffRef (startingIndex_ + static_cast<size_type> (i))
            = coefs_[i](tmax_);
      }
    }
  }
}

#endif //! ROBOPTIM_TRAJECTORY_CONSTRAINTS_OVER_SPLINES_HXX
