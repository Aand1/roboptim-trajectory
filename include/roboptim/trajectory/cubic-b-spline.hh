// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_TRAJECTORY_CUBIC_B_SPLINE_HH
# define ROBOPTIM_TRAJECTORY_CUBIC_B_SPLINE_HH
# include <roboptim/trajectory/sys.hh>

# include <roboptim/trajectory/trajectory.hh>
# include <roboptim/trajectory/fwd.hh>


namespace roboptim
{
  /// \addtogroup roboptim_function
  /// @{

  /** \brief Cubic B-Spline trajectory.
  
      Implement a B-Spline as a trajectory as described below: given 
      \li a number \f$m\geq 4\f$ of control points, 
      \li regularly spaced time points: \f$t_0 < t_1 < \cdots < t_{m}\f$, \f$\forall i\in\{0,...,m-1\}\f$, \f$t_{i+1}-t_i = \Delta t\f$
      \li \f$m\f$ control points \f$ P_0,\cdots, P_{m-1}\f$ in \f$\textbf{R}^n\f$ ,
      the cubic B-spline of control points \f$P_0,\cdots, P_{m-1}\f$ is defined over \f$[t_3,t_{m}]\f$ by
      \f[ B(t) = \sum_{i=0}^{m-1} P_i b_{i,3}(t) \f]
      where basis functions \f$b_{i,3}\f$ are defined by:
      \f{eqnarray*}{
      b_{i,3}(t)=& \frac{(t-t_i)^3}{6\Delta t^3} & \mbox{ if } t_{i} \leq t < t_{i+1} \\
      & \frac{(t-t_i)^2(t_{i+2}-t)+(t-t_i)(t_{i+3}-t)(t-t_{i+1})+(t_{i+4}-t)(t-t_{i+1})^2}{6\Delta t^3}& \mbox{ if } t_{i+1} \leq t < t_{i+2} \\  
      & \frac{(t-t_i)(t_{i+3}-t)^2+(t_{i+4}-t)(t-t_{i+1})(t_{i+3}-t)+(t_{i+4}-t)^2(t-t_{i+2})}{6\Delta t^3}& \mbox{ if } t_{i+2} \leq t < t_{i+3} \\  
      & \frac{(t_{i+4}-t)^3}{6\Delta t^3}& \mbox{ if } t_{i+3} \leq t < t_{i+4} 
      \f}
  */
  class CubicBSpline : public Trajectory<3>
  {
  public:
    /// \brief Instantiate a cubic B-Spline from its definition.
    ///
    /// \param timeRange spline time range: $\f$[t_3,t_n]\f$
    /// \param dimension spline dimension: \f$n\f$
    /// \param parameters vector of parameters defining control points
    /// \param name function title
    ///
    /// Number of control points is inferred from dimension of dimenion of 
    /// parameter vector.
    CubicBSpline (interval_t timeRange, size_type dimension,
		  const vector_t& parameters, 
		  const std::string name = "cubic B-Spline") throw ();

    /// \brief Copy constructor.
    /// \param spline spline that will be copied
    CubicBSpline (const CubicBSpline& spline) throw ();

    virtual ~CubicBSpline () throw ();

    /// \brief Modify spline parameters.
    virtual void setParameters (const vector_t&) throw ();

    virtual jacobian_t variationConfigWrtParam (double t) const throw ();
    virtual jacobian_t variationDerivWrtParam (double t, size_type order)
      const throw ();
    virtual value_type singularPointAtRank (size_type rank) const;
    virtual vector_t derivBeforeSingularPoint (size_type rank, size_type order)
      const;
    virtual vector_t derivAfterSingularPoint (size_type rank, size_type order)
      const;

    ROBOPTIM_IMPLEMENT_CLONE(CubicBSpline)

    virtual Trajectory<derivabilityOrder>* resize (interval_t timeRange)
      const throw ()
    {
      return new CubicBSpline (timeRange, this->outputSize (), this->parameters ());
    }

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream& o) const throw ();

    jacobian_t variationConfigWrtParam (StableTimePoint tp) const throw ();
    jacobian_t variationDerivWrtParam (StableTimePoint tp, size_type order)
      const throw ();

  protected:
    void impl_compute (result_t&, double) const throw ();
    void impl_derivative (gradient_t& g, double x, size_type order)
      const throw ();
    void impl_derivative (gradient_t& g, StableTimePoint, size_type order)
      const throw ();

  private:
    /// \brief Number of control points.
    unsigned int nbp_;
  };

  /// @}

} // end of namespace roboptim.

#endif //! ROBOPTIM_TRAJECTORY_TRAJECTORY_HH
