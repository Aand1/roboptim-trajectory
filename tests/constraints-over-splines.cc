// Copyright (C) 2015 by Benjamin Chr√©tien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <boost/make_shared.hpp>
#include <boost/filesystem/fstream.hpp>

#include <roboptim/core/io.hh>
#include <roboptim/core/util.hh>
#include <roboptim/core/decorator/finite-difference-gradient.hh>
#include <roboptim/core/visualization/matplotlib-function.hh>

#include <roboptim/trajectory/cubic-b-spline.hh>
#include <roboptim/trajectory/b-spline.hh>
#include <roboptim/trajectory/constraints-over-splines.hh>

#include <roboptim/trajectory/visualization/cubic-b-spline-matplotlib.hh>
#include <roboptim/trajectory/visualization/b-spline-matplotlib.hh>

using namespace roboptim;
using namespace roboptim::visualization;
using namespace roboptim::visualization::matplotlib;
using namespace roboptim::trajectory;
using namespace roboptim::trajectory::visualization::matplotlib;

typedef boost::mpl::list<CubicBSpline, BSpline<3> > splinesType_t;

template <typename T>
std::string splineName();

template <>
std::string splineName<roboptim::trajectory::CubicBSpline> ()
{
  return "cubic-b-spline";
}

template <>
std::string splineName<roboptim::trajectory::BSpline<3> > ()
{
  return "b-spline";
}

template <typename T>
std::string getOutputFilename ()
{
  std::string s = "constraints-over-splines-";
  s += splineName<T> ();
  return s;
}

BOOST_FIXTURE_TEST_SUITE (trajectory, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE_TEMPLATE (trajectory_constraints_over_splines,
                               spline_t, splinesType_t)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("constraints-over-splines");

  typedef ConstraintsOverSplines<EigenMatrixDense, spline_t>
    constraintsOverSplines_t;
  typedef typename constraintsOverSplines_t::splines_t splines_t;
  typedef typename spline_t::size_type size_type;

  const size_type n = 10;

  typename spline_t::vector_t x (n);
  x << 10., 5., -3., 5., 4., -2., 8., 4., 1., 0.;

  typename spline_t::interval_t timeRange = CubicBSpline::makeInterval (0., 1.);
  boost::shared_ptr<spline_t> spline = boost::make_shared<spline_t>
                                         (timeRange, 1, x, "spline");
  splines_t splines;
  splines.push_back (spline);

  for (double t = 1e-6; t < 1.; t += 1./7.)
  {
    boost::shared_ptr<constraintsOverSplines_t>
      f = boost::make_shared<constraintsOverSplines_t> (splines, 0, 0, t, n);

    typedef typename finiteDifferenceGradientPolicies::Simple<EigenMatrixDense>
      fdRule_t;
    GenericFiniteDifferenceGradient<EigenMatrixDense, fdRule_t> fd (f);

    (*output)
      << x << "\n"
      << *f << "\n"
      << (*f) (x) << "\n"
      << normalize (toDense (f->gradient (x, 0))) << "\n"
      << normalize (toDense (f->gradient (x, 1))) << "\n"
      << normalize (toDense (f->jacobian (x))) << std::endl;

    // Numerical errors between spline implementations prevent basic string
    // comparisons with FD.
    std::cout
      << fd << "\n"
      << fd (x) << "\n"
      << normalize (toDense (fd.gradient (x, 0)), 1e-6) << "\n"
      << normalize (toDense (fd.gradient (x, 1)), 1e-6) << "\n"
      << normalize (toDense (fd.jacobian (x)), 1e-6) << std::endl;

    BOOST_CHECK (allclose (toDense (f->jacobian (x)),
          toDense (fd.jacobian (x)),
          1e-4, 1e-4));
  }

  Matplotlib matplotlib = Matplotlib::make_matplotlib ();
  matplotlib << plot_spline (*splines[0], 0.005);

  boost::filesystem::ofstream pythonPlot
    (getOutputFilename<spline_t> () + ".py");
  pythonPlot << matplotlib << std::endl;

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
