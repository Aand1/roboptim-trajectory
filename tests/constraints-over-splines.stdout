[10](10,5,-3,5,4,-2,8,4,1,0)
f(t) / t ∈ [0.3000,0.4286] (spline) (differentiable function)
[2](-0.333333,3.83817)
[10](0,0.166667,0.666667,0.166667,0,0,0,0,0,0)
[10](0.1215,0.657167,0.221167,0.000166667,0,0,0,0,0,0)
[2,10]((0,0.166667,0.666667,0.166667,0,0,0,0,0,0), (0.1215,0.657167,0.221167,0.000166667,0,0,0,0,0,0))
f(t) / t ∈ [0.3000,0.4286] (spline) (finite differences):
  Wrapped function: f(t) / t ∈ [0.3000,0.4286] (spline) (differentiable function)
[2](-0.333333,3.83817)
[10](0,0.166669,0.666668,0.166669,0,0,0,0,0,0)
[10](0.1215,0.657166,0.221168,0.000166445,0,0,0,0,0,0)
[2,10]((0,0.166669,0.666668,0.166669,0,0,0,0,0,0), (0.1215,0.657166,0.221168,0.000166445,0,0,0,0,0,0))
